kong/tools/public.lua

```diff
@@ -1,26 +1,71 @@
 local pcall = pcall
 local ngx_log = ngx.log
 local ERR = ngx.ERR
+local str_find = string.find
+
+
+local cjson = require "cjson.safe"
 
 
 local _M = {}
 
 
 do
+  local multipart = require "multipart"
+
+
   local ngx_req_get_post_args = ngx.req.get_post_args
+  local ngx_req_get_body_data = ngx.req.get_body_data
+
+
+  local content_type_map = {
+    [1] = function(content_type) -- multipart
+      return multipart(ngx_req_get_body_data(), content_type):get_all()
+    end,
+
+    [2] = function() -- json
+      local body, err = cjson.decode(ngx_req_get_body_data())
+      if err then
+        ngx_log(ERR, "could not decode JSON body args: ", err)
+        return {}
+      end
+
+      return body
+    end,
 
-  function _M.get_post_args()
-    local ok, res, err = pcall(ngx_req_get_post_args)
+    [3] = function() -- default (encoded form or other)
+      local ok, res, err = pcall(ngx_req_get_post_args)
+      if not ok or err then
+        local msg = res and res or err
+        ngx_log(ERR, "could not get body args: ", msg)
+        return {}
+      end
 
-    if not ok or err then
-      local msg = res and res or err
-      ngx_log(ERR, "could not get body args: ", msg)
-      return {} -- TODO return an immutable table here
+      return res
+    end,
+  }
+
+  function _M.get_body_args()
+    local content_type = ngx.var.http_content_type
+
+    if not content_type or content_type == "" then
+      return {}
+    end
+
+    local map_type
+
+    if str_find(content_type, "multipart/form-data", nil, true) then
+      map_type = 1
+
+    elseif str_find(content_type, "application/json", nil, true) then
+      map_type = 2
+
+    else
+      map_type = 3
```

Comment:
Should we have:
```lua
elseif str_find(content_type, "application/x-www-form-urlencoded", nil, true) then
  map_type = 3
else
  ngx_log(ERR, "could not read body args of content-type: ", content_type)
  return {}
end
```

Since currently, we will attempt `ngx.req.get_post_args()` by default?